plugins/GranularPitchShifter/GranularPitchShifter.h

```diff
@@ -0,0 +1,168 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#ifndef GRANULARPITCHSHIFTER_H
+#define GRANULARPITCHSHIFTER_H
+
+#include "Effect.h"
+#include "GranularPitchShifterControls.h"
+
+#include "BasicFilters.h"
+#include "interpolation.h"
+
+namespace lmms
+{
+
+constexpr float GPS_PREFILTER_BW = 0.96f;// 96% of nyquist
+constexpr double GPS_GLIDE_SNAG_RADIUS = 0.001;
+constexpr int GPS_SAFETY_LATENCY = 3;
+
+// adapted from signalsmith's crossfade approximation:
+// https://signalsmith-audio.co.uk/writing/2021/cheap-energy-crossfade
+inline float cosHalfWindowApprox(float x, float k) {
+	float A = x * (1 - x);
+	float B = A * (1 + k * A);
+	float C = (B + x);
+	return C * C;
+}
+// 1-2 fades between equal-gain and equal-power
+inline float cosWindowApproxK(float p) {
+	return -6.0026608f + p * (6.8773512f - 1.5838104f * p);
+}
+
+class PrefilterLowpass
+{
+public:
+	inline void setCoefs(float sampleRate, float cutoff)
+	{
+		const float g = std::tan(F_PI * cutoff / sampleRate);
+		const float ginv = g / (1.0f + g * (g + 1.414213562));
+		m_g1 = ginv;
+		m_g2 = 2.0f * (g + 1.414213562) * ginv;
+		m_g3 = g * ginv;
+		m_g4 = 2.0f * ginv;
+	}
+
+	inline float process(float input)
+	{
+		const float v1z = m_v1;
+		const float v3 = input + m_v0z - 2.0f * m_v2;
+		m_v1 += m_g1 * v3 - m_g2 * v1z;
+		m_v2 += m_g3 * v3 + m_g4 * v1z;
+		m_v0z = input;
+		return m_v2;
+	}
+
+private:
+	float m_v0z = 0.0f, m_v1 = 0.0f, m_v2 = 0.0f;
+	float m_g1, m_g2, m_g3, m_g4;
+};
+
+class GranularPitchShifterGrain final {
+public:
+	GranularPitchShifterGrain(double grainSpeedL, double grainSpeedR, double phaseSpeedL, double phaseSpeedR, double readPointL, double readPointR) :
+		m_readPoint({readPointL, readPointR}),
+		m_phaseSpeed({phaseSpeedL, phaseSpeedR}),
+		m_grainSpeed({grainSpeedL, grainSpeedR}),
+		m_phase(0) {}
+	std::array<double, 2> m_readPoint;
+	std::array<double, 2> m_phaseSpeed;
+	std::array<double, 2> m_grainSpeed;
+	double m_phase;
+};
+
+class GranularPitchShifterEffect : public Effect
+{
+public:
+	GranularPitchShifterEffect(Model* parent, const Descriptor::SubPluginFeatures::Key* key);
+	~GranularPitchShifterEffect() override = default;
+	bool processAudioBuffer(sampleFrame* buf, const fpp_t frames) override;
+
+	EffectControls* controls() override
+	{
+		return &m_granularpitchshifterControls;
+	}
+	
+	// double index and fraction are required for good quality
+	inline float getHermiteSample(double index, int ch) {
+		const int index_floor = static_cast<int>(index);
+		const double fraction = index - index_floor;
```

Comment:
Can you use const float here instead and avoid the casting at line 130?

Reply:
No, this would very drastically reduce the plugin's audio quality (and wouldn't have any performance benefit since a cast would be required earlier anyway).  The input is already a double, and by waiting until after the integer part is removed to convert it to float, the resulting precision is significantly higher.  Without this, there are extremely ugly audible artifacts, especially when pitch shifting downward, since the float precision limitations would be applied while the index value is quite high.

Reply:
Alright.